packages/react-query/src/HydrationBoundary.tsx

```diff
@@ -19,19 +23,83 @@ export const HydrationBoundary = ({
   queryClient,
 }: HydrationBoundaryProps) => {
   const client = useQueryClient(queryClient)
+  const [hydrationQueue, setHydrationQueue] = React.useState<
+    DehydratedState['queries'] | undefined
+  >()
 
   const optionsRef = React.useRef(options)
   optionsRef.current = options
 
-  // Running hydrate again with the same queries is safe,
-  // it wont overwrite or initialize existing queries,
-  // relying on useMemo here is only a performance optimization.
-  // hydrate can and should be run *during* render here for SSR to work properly
+  // This useMemo is for performance reasons only, everything inside it _must_
+  // be safe to run in every render and code here should be read as "in render".
+  //
+  // This code needs to happen during the render phase, because after initial
+  // SSR, hydration needs to happen _before_ children render. Also, if hydrating
+  // during a transition, we want to hydrate as much as is safe in render so
+  // we can prerender as much as possible.
+  //
+  // For any queries that already exist in the cache, we want to hold back on
+  // hydrating until _after_ the render phase. The reason for this is that during
+  // transitions, we don't want the existing queries and observers to update to
+  // the new data on the current page, only _after_ the transition is committed.
+  // If the transition is aborted, we will have hydrated any _new_ queries, but
+  // we throw away the fresh data for any existing ones to avoid unexpectedly
+  // updating the UI.
   React.useMemo(() => {
     if (state) {
-      hydrate(client, state, optionsRef.current)
+      if (typeof state !== 'object') {
+        return
+      }
```

Comment:
I deleted the previous comment because I was being tired and stupid and gave bad examples.. 😄 (A version of that example has actually happened to me, it was just a bit more involved)

I included this `if` because it's a simple check that stands between crashing the entire app hard or not. As long as it's an object, even if `.queries` isn't available on there we'll fail gracefully and the client can fetch any queries that were meant to be hydrated but weren't. This `if` just takes that graceful fallback to non-object values. Sure, that doesn't protect from `.queries` having a bad format, but that should be more unlikely to happen by accident so I don't think we need runtime validation for that.

On the other hand, it should hopefully still be uncommon that state is truthy but not an object too and crashing hard also gives a better chance to catch it in dev so I don't feel strongly about it, I can go ahead and remove it, can always be added later if it shows up in practice.

Reply:
fair enough 👍